clusterloader2/pkg/measurement/common/service_creation_latency.go

```diff
@@ -0,0 +1,366 @@
+
+	content, err := util.PrettyPrintJSON(measurementutil.LatencyMapToPerfData(serviceCreationLatency))
+	if err != nil {
+		return nil, err
+	}
+	summary := measurement.CreateSummary(serviceCreationLatencyName, "json", content)
+	return []measurement.Summary{summary}, err
+}
+
+func (s *serviceCreationLatencyMeasurement) printLatencies(latencies []measurementutil.LatencyData, header string) {
+	metrics := measurementutil.NewLatencyMetric(latencies)
+	index := len(latencies) - 100
+	if index < 0 {
+		index = 0
+	}
+	klog.Infof("%s: %d %s: %v", s, len(latencies)-index, header, latencies[index:])
+	klog.Infof("%s: perc50: %v, perc90: %v, perc99: %v", s, metrics.Perc50, metrics.Perc90, metrics.Perc99)
+}
+
+type serviceCreationLatencyData struct {
+	Key     string
+	Latency time.Duration
+}
+
+func (s serviceCreationLatencyData) GetLatency() time.Duration {
+	return s.Latency
+}
+
+func (s *serviceCreationLatencyMeasurement) handleObject(oldObj, newObj interface{}) {
+	var oldService *corev1.Service
+	var newService *corev1.Service
+	var ok bool
+	oldService, ok = oldObj.(*corev1.Service)
+	if oldObj != nil && !ok {
+		klog.Errorf("%s: uncastable old object: %v", s, oldObj)
+		return
+	}
+	newService, ok = newObj.(*corev1.Service)
+	if newObj != nil && !ok {
+		klog.Errorf("%s: uncastable new object: %v", s, newObj)
+		return
+	}
+	isEqual := oldService != nil &&
+		newService != nil &&
+		equality.Semantic.DeepEqual(oldService.Spec, newService.Spec) &&
+		equality.Semantic.DeepEqual(oldService.Status, newService.Status)
+	if isEqual {
+		// Skip updates without changes in the spec.
+		return
+	}
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if !s.isRunning {
+		return
+	}
+	if newObj == nil {
+		if err := s.deleteObjectLocked(oldService); err != nil {
+			klog.Errorf("%s: delete checker error: %v", s, err)
+		}
+		return
+	}
+	if err := s.updateObjectLocked(newService); err != nil {
+		klog.Errorf("%s: create checker error: %v", s, err)
+	}
+}
+
+func (s *serviceCreationLatencyMeasurement) deleteObjectLocked(svc *corev1.Service) error {
+	key, err := runtimeobjects.CreateMetaNamespaceKey(svc)
+	if err != nil {
+		return fmt.Errorf("meta key created error: %v", err)
+	}
+	if pc, exists := s.pingCheckers[key]; exists {
+		pc.stop()
+		delete(s.pingCheckers, key)
+	}
+	s.creationTimes.DeleteKey(key)
+	return nil
+}
+
+func (s *serviceCreationLatencyMeasurement) updateObjectLocked(svc *corev1.Service) error {
+	key, err := runtimeobjects.CreateMetaNamespaceKey(svc)
+	if err != nil {
+		return fmt.Errorf("meta key created error: %v", err)
+	}
+	if _, exists := s.creationTimes.Get(key, creatingPhase); !exists {
+		s.creationTimes.Set(key, creatingPhase, time.Now())
+	}
+	if len(svc.Status.LoadBalancer.Ingress) < 1 {
+		return nil
+	}
+	if _, exists := s.creationTimes.Get(key, ipAssigningPhase); !exists {
+		s.creationTimes.Set(key, ipAssigningPhase, time.Now())
+		klog.Infof("%s: %s ipAssigned", s, key)
+	}
+	if _, exists := s.pingCheckers[key]; !exists {
```

Comment:
I think you don't need this if - this should happen exactly when you first observe ingress address.

Reply:
In this implementation this line is the definition of "first observed ingress address".

Reply:
Yes - but the line 287:
"if _, exists := s.creationTimes.Get(key, ipAssigningPhase); !exists"

is also definition of the same thing :)

Reply:
Done.